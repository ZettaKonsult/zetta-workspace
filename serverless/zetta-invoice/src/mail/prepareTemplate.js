/* @flow */

/**
 * @date 2018-02
 */

import type { HTML } from 'types/General';
import type { Invoice, InvoiceSpecification } from 'types/Invoice';
import type { Recipient } from 'types/Recipient';
import Mustache from 'mustache';
import { promisify } from 'util';
import fs from 'fs';

export default async (data: any): HTML => {
  let [template, date] = await Promise.all([
    readTemplateFile(),
    prepareData(data),
  ]);
  console.log(`Prepared template. Translating via Mustache.`);
  return Mustache.to_html(template, date);
};

const readFileAsync = promisify(fs.readFile);
const readTemplateFile = async () =>
  await readFileAsync('./src/template/template.html', 'utf8');

const prepareData = (data: {
  discount: number,
  recipient: Recipient,
  invoice: Invoice,
  tax?: number,
}): InvoiceSpecification => {
  const { discount, recipient, invoice, tax = 1.25 } = data;
  const { companyCustomer } = invoice;

  const id = new Date(invoice.createdAt);
  const timeToPay = new Date(id.getUTCFullYear(), id.getUTCMonth() + 1);

  const { netTotal, taxTotal, total } = calculatePrice({
    price: invoice.price,
    tax,
  });

  return {
    companyCustomer,
    recipient,
    invoice: {
      id: invoice.createdAt,
      createdAt: id.toISOString().split('T')[0],
      timeToPay: timeToPay.toISOString().split('T')[0],
    },
    discount: discount | 0,
    netTotal,
    taxTotal,
    total,
    receiver: recipient.company || recipient.firstName,
  };
};

const calculatePrice = (params: {
  price: number,
  tax: number,
}): { netTotal: number, taxTotal: number, total: number } => {
  const { price, tax } = params;

  const taxTotal = price * tax;
  const netTotal = price - taxTotal;
  const total = netTotal + taxTotal;

  return { netTotal, taxTotal, total };
};
