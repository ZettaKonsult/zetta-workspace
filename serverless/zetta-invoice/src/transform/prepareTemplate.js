/* @flow */

/**
 * @date 2018-02
 */

import type { HTML } from 'types/General';
import type { Invoice, InvoiceSpecification } from 'types/Invoice';
import type { CompanyCustomer, Recipient } from 'types/Recipient';
import Mustache from 'mustache';
import { promisify } from 'util';
import fs from 'fs';

export default async (data: any): HTML => {
  let [tempalte, date] = await Promise.all([
    readTemplateFile(),
    prepareData(data),
  ]);
  return Mustache.to_html(tempalte, date);
};

const readFileAsync = promisify(fs.readFile);
const readTemplateFile = async () =>
  await readFileAsync('./src/template/template.html', 'utf8');

const prepareData = (data: {
  companyCustomer: CompanyCustomer,
  recipient: Recipient,
  invoice: Invoice,
  tax?: number,
}): InvoiceSpecification => {
  const { companyCustomer, recipient, invoice, tax = 1.25 } = data;

  const id = new Date(invoice.createdAt);
  const timeToPay = new Date(id.getUTCFullYear(), id.getUTCMonth() + 1);

  const { netTotal, taxTotal, total } = calculatePrice({
    price: invoice.price,
    tax,
  });

  return {
    companyCustomer,
    recipient,
    invoice: {
      id: invoice.createdAt,
      createdAt: id.toISOString().split('T')[0],
      timeToPay: timeToPay.toISOString().split('T')[0],
    },
    discount: '',
    netTotal,
    taxTotal,
    total,
    receiver: recipient.company || recipient.firstName,
  };
};

const calculatePrice = (params: {
  price: number,
  tax: number,
}): { netTotal: number, taxTotal: number, total: number } => {
  const { price, tax } = params;

  const taxTotal = price * tax;
  const netTotal = price - taxTotal;
  const total = netTotal + taxTotal;

  return { netTotal, taxTotal, total };
};
