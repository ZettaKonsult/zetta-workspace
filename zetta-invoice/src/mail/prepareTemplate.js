/* @flow */

/**
 * @date 2018-02
 */

import type { HTML } from 'types/General';
import type {
  CompanyCustomer,
  Recipient,
  LockedInvoice,
  InvoiceSpecification,
} from 'types/Invoice';
import { to_html } from 'mustache';
import { promisify } from 'util';
import fs from 'fs';

export default async (data: any): HTML => {
  let [template, preparedData] = await Promise.all([
    readTemplateFile(),
    prepareData(data),
  ]);
  return to_html(template, preparedData);
};

const readFileAsync = promisify(fs.readFile);
const readTemplateFile = async () =>
  await readFileAsync('./src/template/template.html', 'utf8');

export const prepareData = (data: {
  discount: number,
  invoice: LockedInvoice,
  companyCustomer: CompanyCustomer,
  recipient: Recipient,
}): Array<InvoiceSpecification> => {
  const { discount, invoice, companyCustomer, recipient } = data;
  const {
    netTotal,
    taxTotal,
    total,
    seperateTaxMap,
    invoiceRows,
  } = invoice.getInvoiceTotal();
  const { createdAt, timeToPay, id } = invoice.getFormatedDateValues();

  const taxTouple = Object.keys(seperateTaxMap).map(key => ({
    tax: convertTaxToPercentage(key),
    value: seperateTaxMap[key],
  }));

  return {
    companyCustomer,
    invoice: {
      id,
      createdAt,
      timeToPay,
    },
    invoiceRows,
    discount: discount | 0,
    netTotal,
    taxTotal,
    total: total,
    recipient,
    taxTouple,
  };
};

export const convertTaxToPercentage = tax => {
  let taxCopy = tax.slice();
  taxCopy = taxCopy.split('.')[1];
  taxCopy = appendTrailingZero(taxCopy);
  taxCopy = removeLeadingZeros(taxCopy);

  return taxCopy + '%';
};
const removeLeadingZeros = string => string.replace(/^0+/, '');
const appendTrailingZero = string =>
  string.length === 1 ? string + '0' : string;
