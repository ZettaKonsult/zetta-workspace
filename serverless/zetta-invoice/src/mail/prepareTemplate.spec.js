/**
 * @date 2018-04-09
 */

import { prepareData } from './prepareTemplate';

describe('Template.', () => {
  it('Prepare data.', () => {
    const data = {
      companyCustomer: {
        id: 'companyCustomerIdA',
        firstName: 'FirstNameA',
        lastName: 'LastNameA',
        city: 'CityA',
        address: 'Road 123AA',
        mobile: '+46761234567A',
        VAT: 'XX123456789101A',
        zipcode: '12345A',
        company: 'Company ABA',
        email: 'firstNameA@company.com',
        defaultTax: 0.2,
        bank: {
          giro: '234-5678',
          name: 'MoneyBankA',
        },
        createdAt: 2345678901,
      },
      recipient: {
        address: 'Road 234A1',
        city: 'RecipientCityA',
        companyCustomerId: 'companyCustomerIdA',
        createdAt: 23456789011,
        email: 'firstNameA@recipient.com',
        firstName: 'RecipientFirstA',
        id: 'recipientIdA',
        lastName: 'RecipientLastA',
        mobile: '+467623456781',
        ssn: '12345678901',
        zipcode: '123451',
      },
      invoice: {
        id: 'invoiceId1A',
        companyCustomerId: 'companyCustomerIdA',
        createdAt: 3456789012,
        invoiceRows: [
          {
            price: 123,
            unit: 10,
            tax: 0.01,
            description: 'Aan',
          },
        ],
        itemStatus: {
          id: 'itemStatusId1A',
          invoiceId: 'invoiceId1A',
          createdAt: 3456789012,
          itemStatus: 'pending',
        },
        locked: true,
      },
      discount: 0,
    };

    const result = prepareData(data);

    expect(result).toMatchSnapshot();
  });
});
