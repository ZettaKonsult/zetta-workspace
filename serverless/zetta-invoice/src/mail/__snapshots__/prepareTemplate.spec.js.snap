// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template. Prepare data. 1`] = `
Object {
  "companyCustomer": Object {
    "VAT": "XX123456789101A",
    "address": "Road 123AA",
    "bank": Object {
      "giro": "234-5678",
      "name": "MoneyBankA",
    },
    "city": "CityA",
    "company": "Company ABA",
    "createdAt": 2345678901,
    "defaultTax": 0.2,
    "email": "firstNameA@company.com",
    "firstName": "FirstNameA",
    "id": "companyCustomerIdA",
    "lastName": "LastNameA",
    "mobile": "+46761234567A",
    "zipcode": "12345A",
  },
  "discount": 0,
  "invoice": Object {
    "createdAt": "1970-02-10",
    "id": 3456789012,
    "timeToPay": "1970-02-28",
  },
  "invoiceRows": Array [
    Object {
      "description": "Aan",
      "price": 123,
      "tax": 0.01,
      "total": 1230,
      "unit": 10,
    },
  ],
  "netTotal": 1230,
  "recipient": Object {
    "address": "Road 234A1",
    "city": "RecipientCityA",
    "companyCustomerId": "companyCustomerIdA",
    "createdAt": 23456789011,
    "email": "firstNameA@recipient.com",
    "firstName": "RecipientFirstA",
    "id": "recipientIdA",
    "lastName": "RecipientLastA",
    "mobile": "+467623456781",
    "ssn": "12345678901",
    "zipcode": "123451",
  },
  "taxTotal": 12.3,
  "total": 1242.3,
}
`;
